### ISSUE TITLE
Creating Cell Flow Graph Without Ipython

### ISSUE THREAD

**Issue by Red-Giuliano** [ipyflow/ipyflow#134]
Hello!

Let me start by saying thank you so much for this! I've read one of the papers behind this and I'm impressed by the depth and breadth of the project.

My question is: would it be possible to use this library without using ipython? For example if I have a dictionary of different code snippets (much like a jupyter notebook) and I would like to parse the dependencies between them is there any easy way to do this using subrocess, exec or the exec function provided by the pyccolo library? 

Thanks in advance for your time and help!

**Comment by smacke** [comment]
Hi there,

If you don't want to execute with ipython, it's possible in theory, but not supported right now (certainly not with just pyccolo). Basically you'd have to trace through all the code paths from `IPyflowKernel.do_execute` (including AST transformers running implicitly) and break all of that out. It would be very messy and a lot of work.

If executing the code in order to determine the dependencies is OK for your use case, is there a reason it can't go through ipython?

**Comment by Red-Giuliano** [comment]
Hey thanks so much for the fast reply! I don't know if it's worth going into the details for why I don't want to use ipython, but we would much prefer using exec. Are you planning on supporting this sort of functionality at any point in the future? You mention it is possible in theory... it sounds tedious but how long do you think it would take in practice? I think I would be willing to try to take a stab with a bit more guidance.

I had been reading over the codebase and I figured that a lot of the work for the package was being done in core/ipyflow/kernel/kernel.py. I'm guessing this is what would mainly have to be rewritten. Could you elaborate on what you mean when you say "including AST transformers running implicitly"? Thanks so much for the time and help with all this.

**Comment by smacke** [comment]
> Are you planning on supporting this sort of functionality at any point in the future?

No concrete plans.

> You mention it is possible in theory... it sounds tedious but how long do you think it would take in practice?

For me, probably a few weeks of effort (which I unfortunately don't have). For anybody else, probably much longer. There's a lot of tech debt and poorly documented parts of the code base. If you want to take it on, godspeed. I can't merge any PRs unless all the tests are passing though (you can look at the Makefile or ci.yml to see how to run these).

> I had been reading over the codebase and I figured that a lot of the work for the package was being done in core/ipyflow/kernel/kernel.py. I'm guessing this is what would mainly have to be rewritten.

Unfortunately I think there is a lot of messy and coupled supporting code in `flow.py`, `ipyflow_tracer.py`, and maybe some other places.

> Could you elaborate on what you mean when you say "including AST transformers running implicitly"?

IPython allows you to register AST transformers that run on the code text before it gets turned into bytecode. Normally we'd let pyccolo take care of this part, but I wasn't sure if we could implement the kernel's `do_execute` in terms of `pyc.exec` directly -- I found it easier to call the superclass `do_execute` so that it could handle all of the ipython protocol stuff. (This is something you'd probably need to figure out if you decide to decouple ipyflow from ipython.) The part where we actually do this AST transformer registration in ipyflow is here: https://github.com/ipyflow/ipyflow/blob/master/core/ipyflow/kernel/kernel.py#L349

> I don't know if it's worth going into the details for why I don't want to use ipython, but I would much prefer using exec.

Might still be worth going through these -- I'm having trouble imagining a use case where it makes more sense to go through all the effort of rewriting ipyflow versus just threading the executions through ipython running with the ipyflow kernel, and there may be a way to fit the latter approach to your particular use case that may have been overlooked.

**Comment by Red-Giuliano** [comment]
Thanks so much for being so thorough. I didn't realize how large the scope of work would be to do something like that... The reason I am using exec is that I want to restrict the access to global variables that each cell has. Ipython has one state for the entire notebook and all cells have access to this state. I want to make sure that this is not the case as I am not planning on running all of these code snippets in the same process. I don't want subsequent cells to necessarily have access to the outputs of previous cells. In order to restrict the variables for each execution I am passing a dictionary as the globals in the exec() function and then I am storing the outputs of each code snippet to be inspected later. Hopefully that makes sense... the use case is a bit peculiar but this package seems to do a lot of stuff I've been trying to code up for myself. If you can think of a way of passing globals to ipython's do_execute and then storing the outputs on a cell by cell basis let me know!

**Comment by smacke** [comment]
> I don't want subsequent cells to necessarily have access to the outputs of previous cells. In order to restrict the variables for each execution I am passing a dictionary as the globals in the exec() function and then I am storing the outputs of each code snippet to be inspected later.

You may have luck overriding the values in `get_ipython().user_ns` to only those variables that you want a given cell to see:

```python
from IPython import get_ipython
user_ns = get_ipython().user_ns
prev_vals = user_ns.copy()
user_ns.clear()
for v in allowed_vars:
    user_ns[v] = prev_vals[v]
```

**Comment by Red-Giuliano** [comment]
Ok this is interesting I'm reading through some of the docs now... it seems to be the ipython equivalent to globals if I understand correctly. So imagine I have a dictionary with code and allowed_vars for each cell... is there an easy way to return the graph generated by ipyflow? Sorry for all the questions this has been incredibly helpful already.

**Comment by smacke** [comment]
You need a way to associate each cell with some id, so that you can query ipyflow's cell model for each cell's parents / children. I think the default is to use execution counter minus 1 as the id. Example:

```
jupyter console --kernel ipyflow

In[1]: x = 0

In[2]: y = x + 1

In[3]: from ipyflow import cells

In[4]: cells().from_id(0).text
Out[4]: 'x = 0'

In[5]: from ipyflow.slicing.context import dynamic_slicing_context, static_slicing_context

In [6]: with dynamic_slicing_context():
   ...:     print(cells().from_id(0).children)
   ...:
{1: {<x>}}
```

The reason for the context managers is because, in ipyflow, cells can have static parents, static children, dynamic parents, and dynamic children. Static parents / children are inferred just by looking at the code (i.e. without running it), while dynamic parents / children are determined from pyccolo instrumentation (i.e. by actually running it). Depending on which context manager you use, you'll either get the static or dynamic edges. Generally the dynamic edges are higher quality / more accurate. If you want to be thorough you can check both. In the example above, we can see that the 'x = 0' cell has as a dynamic dependent the 'y = x + 1' cell, and it's the 'x' symbol that introduces the dependency.

**Comment by Red-Giuliano** [comment]
That makes a lot of sense as far as why you need to execute the code. I have the code snippets in a dictionary already with a unique id. Is there any way to achieve what you did above programmatically instead of running manually?

**Comment by smacke** [comment]
Yes I think that's the way to go for now unfortunately. The cell magic was something I wrote back in early 2020 when I had no idea what I was doing; the proper way should have been to inherit from `InteractiveShell` (created #137 to track).

**Comment by smacke** [comment]
To create IPyflow variants of all the various subclasses of `InteractiveShell`, it seems like the cleanest way (but definitely not the easiest to understand) is to do fancy metaclass stuff to patch `IPyflowInteractiveShell` into the mro. Toy example:

```python
from IPython.terminal.embed import InteractiveShellEmbed
from IPython.terminal.interactiveshell import InteractiveShell
from traitlets import MetaHasTraits


def make_mro_inserter_metaclass(old_class, new_class):
	class MetaMroInserter(MetaHasTraits):
		def mro(cls):
			ret = []
			for clazz in super().mro():
				if clazz is old_class:
					ret.append(new_class)
				ret.append(clazz)
			return ret
	return MetaMroInserter


class MyInteractiveShell(InteractiveShell):
	def run_cell(self, *args, **kwargs):
		print("run cell")
		return super().run_cell(*args, **kwargs)


class MyInteractiveShellEmbed(InteractiveShellEmbed, metaclass=make_mro_inserter_metaclass(InteractiveShell, MyInteractiveShell)):
	pass


ipshell = MyInteractiveShellEmbed()
ipshell.run_cell("x = 0")  # prints 'run cell' before running the cell
```

**Comment by smacke** [comment]
Oh, if you want to get the graph, then yes you probably want to write a helper to traverse parents for all of the cells you can get from cells(). (`[cells().at_timestamp(ts) for ts in range(cells().exec_counter())]`)

**Comment by Red-Giuliano** [comment]
Thank you for all the help with this... other priorities have come up for me right now so I am continuing to use a static analysis solution for the time being but I am planning on returning to this. I'll open another issue then.
